### PR TITLE
avoid unnecessary workers with sequential `CombinedLoader `

### LINKED ISSUE #17641: potentially substantial overhead starting all dataloaders in CombinedLoader
### Bug description

When training with multiple validation dataloaders, Lightning simultaneously initializes all dataloaders causing each one to initialize a full set of workers and begin prefetching potentially introducing substantial overhead. For example, if there are 100 dataloaders and each one is set to `num_workers=32` then Lightning will initialize all 100 dataloaders during the validation process causing 3200 workers to spawn and a lot of batches to prefetch.

### What version are you seeing the problem on?

v2.0, master

### How to reproduce the bug

_No response_

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

_No response_

cc @justusschock @awaelchli @carmocca

### PR DESCRIPTION
**Pull request by mukhery**
## What does this PR do?
This PR relates to behavior exhibited when training with multiple validation dataloaders. In this scenario, Lightning wraps the multiple dataloaders with a `CombinedLoader`. This `CombinedLoader` currently initializes all of the dataloaders causing all of their workers to start up and prefetch data potentially introducing substantial overhead.

Fixes #17641

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/pytorch/utilities/combined_loader.py [hunk 1]
```diff
@@ -131,19 +131,24 @@ def __next__(self) -> Tuple[Any, int, int]:
...
         return self
 
     def reset(self) -> None:
         super().reset()
         self._iterator_idx = 0
         self._idx = 0
 
+    def _load_current_iterator(self) -> None:
+        # Load a single DataLoader, prevents multiple sets of workers from starting unnecessarily
+        self.iterators = [iter(iterable) for idx, iterable in enumerate(self.iterables) if idx == self._iterator_idx]
```
Comment (carmocca) [inline src/lightning/pytorch/utilities/combined_loader.py:146]:
```suggestion
        self.iterators = [iter(self.iterables[self._iterator_idx])]
```

**Review by carmocca — COMMENTED** [review]
Good catch! Can you add a test in tests/tests_pytorch/utilities/test_combined_loader.py. Also a "Fixed" entry in src/lightning/pytorch/CHANGELOG.md

**Comment by mukhery** [comment]
> Good catch! Can you add a test in tests/tests_pytorch/utilities/test_combined_loader.py. Also a "Fixed" entry in src/lightning/pytorch/CHANGELOG.md

Alright, I think I've addressed these.

**Review by Borda — APPROVED** [review]

File: tests/tests_pytorch/loops/test_loops.py [hunk 2]
```diff
@@ -844,9 +844,9 @@ def _get_iterator(self):
             # iterable check
             0,
             # epoch ends
+            0,
             1,
             # teardown
-            1,
```
Comment (carmocca) [inline tests/tests_pytorch/loops/test_loops.py:849]: Is this not getting called on `teardown` anymore because `_load_current_iterator` will set `self.iterators = []` and trigger garbage collection and thus iterator shutdown?

File: tests/tests_pytorch/loops/test_loops.py [hunk 3]
```diff
@@ -844,9 +844,9 @@ def _get_iterator(self):
             # iterable check
             0,
             # epoch ends
+            0,
             1,
             # teardown
```
Comment (carmocca) [inline tests/tests_pytorch/loops/test_loops.py:849]: dangling comment

```suggestion
```

File: tests/tests_pytorch/loops/test_loops.py [hunk 4]
```diff
@@ -855,9 +855,9 @@ def _get_iterator(self):
             # iterable check
             0,
             # epoch ends
+            0,
             1,
             2,
             # teardown
```
Comment (carmocca) [inline tests/tests_pytorch/loops/test_loops.py:861]:
```suggestion
```

File: tests/tests_pytorch/utilities/test_combined_loader.py [hunk 5]
```diff
@@ -305,6 +305,41 @@ def test_combined_loader_sequence_iterable_dataset(mode, use_multiple_dataloader
...
+            super()._shutdown_workers()
+
+    loaders = [
+        TestDataLoader(range(10), batch_size=2, num_workers=0),
+        TestDataLoader(range(20), batch_size=2, num_workers=0),
+    ]
+    combined_loader = CombinedLoader(loaders, mode)
+
+    for idx, item in enumerate(combined_loader):
+        break
```
Comment (carmocca) [inline tests/tests_pytorch/utilities/test_combined_loader.py:332]: Did you use enumerate for a specific reason?

Comment (mukhery) [reply tests/tests_pytorch/utilities/test_combined_loader.py]: nope, copy/paste from other tests

File: tests/tests_pytorch/loops/test_loops.py (hunk 2 above)
Comment (mukhery) [reply tests/tests_pytorch/loops/test_loops.py:849]: Yeah! Should be that `__next__` gets called after the last item in the last dataloader -> raises `StopIteration` -> catch and trigger `_use_next_iterator` -> set `self.iterators = []` -> workers get shutdown -> raise `StopIteration` -> Lightning now knows that we ran out of data in the epoch

**Review by carmocca — APPROVED** [review]
Thanks for fixing this!

**Review by awaelchli — APPROVED** [review]
